README.md
# Monitoring and controlling Midea-like air conditioners

[![Build](https://github.com/reneklootwijk/node-mideahvac/workflows/build/badge.svg)](https://github.com/reneklootwijk/node-mideahvac/actions)
[![Coverage Status](https://coveralls.io/repos/github/reneklootwijk/node-mideahvac/badge.svg?branch=master)](https://coveralls.io/github/reneklootwijk/node-mideahvac?branch=master)
[![npm](https://img.shields.io/npm/v/node-mideahvac)](https://www.npmjs.com/package/node-mideahvac)

This module enables the monitoring and controlling of 'Midea'-like airconditioners. The remote control functionality via WiFi provided by various vendors of air conditioners, either as default or optional feature, is using the same type of interface and the Midea cloud. Examples of vendors using this interface are:

* Midea
* Qlima
* Artel
* Carrier

The WiFi interface is provided by a dongle, called WiFi SmartKey, either connected to an USB type-A connector or a JST-HX type of connector. This dongle wraps the UART protocol used to communicate with the AC unit with a layer for authentication and encryption for communication with a mobile app via the Midea cloud or directly via a local LAN connection. However, it turned out the dongle is just connected to a serial interface (TTL level) of the AC unit. This means an alternative is to hook up directly to this serial interface and bypass the cloud, authentication and encryption stuff, for instance using an ESP8266 (see [ADAPTER.md](./ADAPTER.md)).

In the current version of this module communication using the WiFi SmartKey and the Midea Cloud and communication via a TCP serial bridge connected to the UART port is supported.
Direct connection to the WiFi SmartKey interface via a local LAN connection will be added later.

## References and sources

The knowledge to create the logic to monitor and control the Midea-'like' AC units was gathered by reverse engineering Android applications from various vendors, analyzing the UART protocol between an Artel unit and an SK103 SmartKey, documents found on the Internet, primarily in Chinese and the work from:

* Mac Zhou: <https://github.com/mac-zhou/midea-msmart>
* NeoAcheron: <https://github.com/NeoAcheron/midea-ac-py>

## Status

There are still a lots of unknowns in the protocol, either because they are really unknown or because I was not able to test them on the AC units I have myself (Artel). Any pull requests to improve and enhance the module are welcome.

## Installation

```bash
npm install node-mideahvac
```

## Usage

First create an appliance instance by specifying the following parameters:

* `communicationMethod`, this must be either 'mideacloud' or 'serialbridge'
* `host`, in case of the mideacloud method this will be the address of the cloud service which defaults to mapp.appsmb.com. In case of the serialbridge method this is the address of the serialbridge.
* `port`, in case of the mideacloud method this will be the port of the cloud service which defaults to 443. In case of the serialbridge method this is the port of the serialbridge which defaults to 23.
* `appKey`, the appKey authenticates the application when using the Midea cloud. Optionally you can specify your own, but it defaults to a hardcoded appKey.
* `uid`, the user Id to logon to the Midea Cloud
* `password`, the password to login into the Midea Cloud
* `deviceId`, the id of the device to address via the Midea Cloud

For each AC unit to be monitored and controlled an appliance must be created. When multiple appliances are configured using the same uid for the Midea Cloud, the connection to the cloud will be shared between these appliances. However, it is also possible to monitor and control appliances linked to different uid's.

An example of creating an appliance using the Midea Cloud:

```javascript
const appliances = require('node-mideahvac')

var options = {
    communicationMethod: 'mideacloud',
    uid: <your email address>,
    password: <your password>,
    deviceId: <the id of the device to control>
}

var ac = appliances.createAppliance(options)
```

An example of creating an appliance using a TCP serial bridge:

```javascript
const appliances = require('node-mideahvac')

var options = {
    communicationMethod: 'serialbridge',
    host: '192.168.10.34',
    port: 23
}

var ac = appliances.createAppliance(options)
```

### Methods

All methods return a promise. An application should first use the `initialize` method to obtain the current status of the unit.

The following methods are provided:

* `getCapabilities(retry)`, this method requests the AC unit for its supported capabilities (B5 query). The retry parameter indicates whether or not the command must be retried when a retryable error occurs (default = false). The promise resolves to a JSON object containing the following capabilities (how many capabilities are actually reported dependents on what your unit reports):

| Capability | Type | Description |
| --- | --- | --- |
| autoMode | boolean | automatically select cool or heat mode |
| autoAdjustDownTemp | number | a specified setpoint under this temperature will be set to this temperature in auto mode |
| autoAdjustUpTemp | number | a specified setpoint above this temperature will be set to this temperature in auto mode |
| coolMode | boolean | cool mode |
| coolAdjustDownTemp | number | a specified setpoint under this temperature will be set to this temperature in cool mode |
| coolAdjustUpTemp | number | a specified setpoint above this temperature will be set to this temperature in cool mode |
| dryMode | boolean | dry mode |
| ecoMode | boolean | eco mode |
| frostProtectionMode | boolean | Maintain temperature at 8C to protect against frost|
| electricAuxHeating | boolean | ? |
| hasAutoClearHumidity | boolean | Intelligent dehumification?? the desired humidity level is determined automatically|
| hasAvoidPeople | boolean | ? |
| hasBlowingPeople | boolean | ? |
| hasBreeze | boolean | ? |
| hasHandClearHumidity | boolean | Manual dehumification?? the desired humidity level can be set |
| hasNoWindFeel | boolean | ? |
| hasNoWindSpeed | boolean | ? |
| hasSelfClean | boolean | ?When the unit is turned off the fan continues to work for another 10 ~ 15 minutes by ventilating and drying the heat exchanger |
| heatMode | boolean | heat mode |
| heatAdjustDownTemp | number | a specified setpoint under this temperature will be set to this temperature in heat mode |
| heatAdjustUpTemp| number | a specified setpoint above this temperature will be set to this temperature in heat mode |
| horizontalSwingMode | boolean | swing up and down |
| isHavePoint | boolean | ? |
| leftNum | boolean | ? |
| lightType | boolean | ? |
| mutilTemp | boolean | ? |
| nestCheck | boolean | ? |
| nestNeedChange | boolean | ? |
| powerCal | boolean | ? |
| powerCalSetting | boolean | ? |
| selfcheck | boolean | ? |
| specialEco | boolean | ? |
| strongCoolMode | boolean | turbo function in cool mode |
| strongHeatMode | boolean | turbo function in heat mode |
| unitChangeable | boolean | ? |
| verticalSwingMode | boolean | swing left and right |

* `getPowerUsage(retry, emitUpdates)`, this method requests the current power usage of the unit (specific 41 command). The retry parameter indicates whether the command must be retried in case of a retryable error occurs (default = false) and the emitUpdates parameter indicates whether or not the status-update event must be emitted when updates are detected (default = true). The promise resolves to a JSON object containing the property values when successful. The following properties are reported:

| Property | Type | Description |
| --- | --- | --- |
| powerUsage | number | Power usage in kWh |

* `getStatus(retry, emitUpdates)`, this method requests the current status of the unit (41 command). The retry parameter indicates whether the command must be retried in case of a retryable error occurs (default = false) and the emitUpdates parameter indicates whether or not the status-update event must be emitted when updates are detected (default = true). The promise resolves to a JSON object containing the property values when successful. The following properties are reported:

| Property | Type | Description |
| --- | --- | --- |
| alarmSleep | boolean | ? |
| braceletHomeAwayMode | boolean | ? only reported by certain units |
| braceletSleepMode | boolean | ? only reported by certain units |
| catchCold | boolean | ? |
| childSleepMode | boolean | ? |
| coolWindMode | boolean | ? |
| cosySleepMode | number | ? |
| downWindControl | boolean | ? |
| downWindControlLR | boolean | ? |
| dryClean | boolean | ? |
| dustFull | boolean | ? |
| ecoModeActive | boolean | flag to indicate eco mode is active|
| ecoSleepRunningHours | number | ? |
| ecoSleepRunningMinutes | number | ? |
| ecoSleepRunningSeconds | number | ? |
| errorCode | object | error code and description when the unit is in error |
| fanSpeed | object | 20: silent, 40: low, 60: medium, 80: high, 102: auto |
| fastCheckActive | boolean | ? |
| frostProtectionModeActive | boolean | flag indicating frost protection is active (only in heat mode) |
| mode | object| 1: auto, 2: cool,3: dry, 4: heat, 5: fanonly |
| horizontalSwingActive | boolean | horizontal swinger on or off |
| humiditySetpoint | number | |
| indoorTemperature | number | temperature measured by the indoor unit |
| inError | boolean | Flag to determine if the unit is in error |
| keepWarm | boolean | ? only reported by certain units || lightClass | number | ? |
| lowFrequencyFan | boolean | ? |
| naturalWindModeActive | boolean | ? |
| nightLight | boolean | ? |
| offTimerActive | boolean | flag to indicate the timer has been set to turn the unit on |
| offTimerHours | number | the number of hours before the timer expires |
| offTimerMinutes | number | the number of minutes before the timer expires |
| onTimerActive | boolean | flag to indicate the timer has been set to turn the unit on |
| onTimerHours | number | the number of hours before the timer expires |
| onTimerMinutes | number | the number of minutes before the timer expires |
| outdoorTemperature | number | temperature measured by the outdoor unit |
| peakValleyMode | boolean | ? |
| personalFeeling | boolean | ? |
| pmvMode | number | redicted Mean Vote Model? |
| powerOn | boolean | Is the unit turned on |
| powerSave | boolean | ? |
| ptcHeaterActive | boolean | ? |
| purifyingModeActive | boolean | ? |
| resumeActive | boolean | After a power failure the unit resumes the previous state |
| selfPersonalFeeling | boolean | ? |
| setpoint | number | The setpoint temperature |
| settingTemperature2 | boolean | ? |
| sleepModeActive |  boolean | flag indicating sleep mode is active or not | strong | boolean | |
| temperatureUnit | object | unit in which the temperature is reported, 0: celcius, 1: fahrenheit |
| timerMode | object | 0: absolute time, 1: relative time |
| turboModeActive | boolean | flag indicating turbo mode is active or not |
| ventilation | boolean | ? |
| verticalSwingActive | boolean | vertical swinger on or off |
| windBlowing | boolean | ? only reported by certain units |
| wiseEye | boolean | ? |

* `initialize()`, the initialize command will connect to the unit and use the `getCapability` and `getStatus` initialize the unit.  The promise will resolve to a nested JSON object containing a status object with all properties and their values and a capabilities object containing all capabilities reported by the unit.

* `setStatus(properties, retry, emitUpdates)`, this method must be used to change the status of the unit. The properties parameter is an object containing all the properties and their values that need to be changed. The retry parameter indicates whether the command must be retried in case of a retryable error occurs (default = false) and the emitUpdates parameter indicates whether or not the status-update event must be emitted when updates are detected (default = true). The response is a JSON object containing all the properties with their values.

In this version the following properties can be set:

 Property | Type | Possible values | Description |
| --- | --- | --- | --- |
| beep | boolean | true, false | en/disable a beep as feedback |
| fanSpeed | string | low, medium, high, auto | set the fan speed |
| frostProtectionModeActive | boolean | true, false | turn frost protection mode (min. temperature is 8°C) on/off. This is only supported in heat mode |
| horizontalSwingActive | boolean | true, false | turn the horizontal swinger on/off |
| mode | string | cool, heat, fanonly, dry, auto | set the operational mode |
| powerOn | boolean | true, false | power the unit on/off |
| setpoint | number | 16 - 31 / 60 - 88| set the desired temperature in °C or °F|
| sleepModeActive | boolean | true, false | turn the sleep mode on/off |
| temperatureUnit | string | fahrenheit, celcius | set the temperature unit to fahrenheit/celcius |
| turboModeActive | boolean | true, false | turn turbo mode on/off |
| verticalSwingMode | boolean | true, false | turn the vertical swinger on/off |

Example:
Turn the unit on, set the setpoint to 24°C, retry until successful and emit the `status-update` event to report the new property values.

```javascript
ac.setStatus({ powerOn: true, setpoint: 24 }, true, true)
.catch(error => {
    console.log(error.message)
})
```

### Events

The following events are emitted:

* `connected`, this event is emitted when using the serialbridge communication method and the connection to the bridge has been established.

* `disconnected`,this event is emitted when using the serialbridge communication method and the connection is disconnected.

* `initialized`, this event is emitted when the `initialize` method has completed. The data contains the same nested JSON object as described with the `initialize` method.

* `status-update`, this event is emitted when the value of one or more properties is updated. The data is a JSON object containing all updated properties with their values.

# Logging

The Winston module is used for logging. You can configure the logging by creating your own logger, e.g.:

```javascript
const logger = require('winston')

logger.remove(logger.transports.Console)
logger.add(new logger.transports.Console({
  format: logger.format.combine(
    logger.format.timestamp(),
    logger.format.colorize(),
    logger.format.printf(event => {
      return `${event.timestamp} ${event.level}: ${event.message}`
    })
  ),
  level: 'silly'
}))
```

By specifying the level you can control the amount of logging generated.

More information on Winston can be found here: <https://www.npmjs.com/package/winston>

# Observations

As a test I monitored my 4 identical AC units (Artel) for a couple of days 24hr/day. 1 unit was monitored via a serialbridge connection the others via the Midea Cloud. The units monitored via the Midea Cloud reported sometimes wrong values for various properties (e.g. powerOn, hortizontalSwingActive, setpoint) while the unit monitored via the serialbridge did not show this behaviour. Also the Midea Cloud regularly returned an error (timeout or system error) or was unavailable completely for hours. This makes me conclude the Midea Cloud is unreliable.

# To do

* Add support for the direct LAN communication via the WiFi SmartKey


CHANGELOG.md
# Changelog

## 0.1.19

* Fixed sending non 0x02 and 0x03 message types, these should not have a crc added.

## 0.1.18

* Added support for reporting the actual power usage of the unit (getPowerUsage method and C1 parser).

## 0.1.14

* Turns out the sendNetworkStatusNotification must be send once each 2 minutes to prevent the WiFi symbol from disappearing.

## 0.1.13

* Added a parser for the C8 response used by the DeHumidifier appliance (thanks to Sören Beye)

## 0.1.12

* When parsing the C0 response the turbo property is now set when the strong and/or the turbo bit has been set
* Changed the setStatus command to also set the stronger bit when the turbo mode must be set
* The function to add the header now also takes a parameter to specify the agreement version
* Changed the getCapabilities command (no impact on response)
* Changed the getStatus command (no impact on response)
* Fixed bug in sendNetworkStatusNotification (AP mode changed into client mode) this prevents the WiFi symbol from disappearing after approx. 20s
* Continue initialization even when a unit does not support the sendNetworkStatusNotification and/or getCapabilities commands
* Updated unit tests

## 0.1.11

* Fixed bug in setting the horizontalSwingActive property.


## 0.1.10

* Send network status notification message during initialization of the serialbridge communication method. The result is the WiFi symbol will be shown in the dispay of the AC unit.


lib/ac.js
'use strict'

const EventEmitter = require('events').EventEmitter
const logger = require('winston')

const { createCommand } = require('./ac_common')
const errors = require('./errors')
const { parse } = require('./parsers')
const reporter = require('./reporter')

// Add a transport as fall back when no parent logger has been initialized
// to prevent the error: "Attempt to write logs with no transports"
logger.add(new logger.transports.Console({
  level: 'none'
}))

module.exports = class extends EventEmitter {
  constructor (options = {}) {
    super()

    this._initialized = false

    this._messageId = 1

    // Capabilities
    this.capabilities = {}

    // Status
    this.status = {}
  }

  _updateStatus (properties) {
    var self = this
    const updates = {}

    logger.silly(`AC._updateStatus: Entering with ${JSON.stringify(properties)}`)

    for (const property in properties) {
      if (self.status[property] !== properties[property]) {
        logger.debug(`AC._updateStatus: Update ${property} from ${self.status[property]} to ${properties[property]}`)

        self.status[property] = properties[property]
        updates[property] = properties[property]
      }
    }

    return updates
  }

  getCapabilities (retry = false) {
    var self = this

    let cmd = Buffer.from([
      0xB5, 0x01, 0x01, 0x01
    ])

    cmd = createCommand(cmd, 0x03)

    return new Promise((resolve, reject) => {
      self._request(cmd, 'getCapabilities', retry)
        .then(response => {
          // Check this is the correct response type
          if (response[10] !== 0xB5) {
            return reject(new Error('Invalid response'))
          }

          const parsedData = parse(response)

          // Persist capability list in-memory
          self._capabilities = parsedData

          resolve(parsedData)
        })
        .catch(error => {
          reject(error)
        })
    })
  }

  // This command is only supported in serialbridge mode, the SmartKey will not
  // forward it
  getElectronicId () {
    var self = this

    let cmd = Buffer.from([0x00])

    cmd = createCommand(cmd, 0x07, 0xFF, false)

    return new Promise((resolve, reject) => {
      self._request(cmd, 'getElectronicId')
        .then(response => {
          resolve(response)
        })
        .catch(error => {
          reject(error)
        })
    })
  }

  getPowerUsage (retry = false, emitUpdates = true) {
    var self = this
    let cmd = Buffer.from([
      0x41, 0x21, 0x01, 0x44, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x04
    ])

    cmd = createCommand(cmd, 0x03)

    return new Promise((resolve, reject) => {
      // Send the command
      self._request(cmd, 'getPowerUsage', retry)
        .then(response => {
          // Check this is the correct response type
          if (response[10] !== 0xC1) {
            return reject(new Error('Invalid response'))
          }

          const parsedData = parse(response)

          // Update in-memory state
          const updates = self._updateStatus(parsedData)

          if (emitUpdates && Object.keys(updates).length) {
            self.emit('status-update', reporter(updates))
          }

          resolve(reporter(parsedData))
        })
        .catch(error => {
          reject(error)
        })
    })
  }

  getStatus (retry = false, emitUpdates = true) {
    var self = this

    let cmd = Buffer.from([
      0x41, 0x81, 0x00, 0xFF, 0x03, 0xFF,
      0x00, 0x02, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x00, 0x03
    ])

    cmd = createCommand(cmd, 0x03)

    return new Promise((resolve, reject) => {
      // Send the command
      self._request(cmd, 'getStatus', retry)
        .then(response => {
          // Check this is the correct response type
          if (response[10] !== 0xC0) {
            return reject(new Error('Invalid response'))
          }

          const parsedData = parse(response)

          // Update in-memory state
          const updates = self._updateStatus(parsedData)

          if (emitUpdates && Object.keys(updates).length) {
            self.emit('status-update', reporter(updates))
          }

          resolve(reporter(parsedData))
        })
        .catch(error => {
          reject(error)
        })
    })
  }

  // This command is only supported in serialbridge mode, the SmartKey will not
  // forward it
  // This message is send when the status of the network connection changes
  // or on response on a request of the AC unit (message type 0x63)
  // Sending this update show the WiFi symbol on the display of the unit
  sendNetworkStatusNotification () {
    var self = this
    let cmd = Buffer.alloc(20)

    // Byte 0: Module type:
    // - 0x00 RF module
    // - 0x01 WiFi module
    cmd[0] = 0x01

    // Byte 1: WiFi module working mode:
    // - 0x01 Client mode
    // - 0x02 Configuration mode
    // - 0x03 AP mode
    cmd[1] = 0x01

    // Byte 2: WiFi signal strength
    // - 0x00 No signal
    // - 0x01 Weak
    // - 0x02 Low
    // - 0x03 Medium
    // - 0x04 Strong
    // - 0xFF WiFi is not supported
    cmd[2] = 0x04

    // Byte 3-6: IP address of client in reverse order
    let digits = {}
    if (self._connection && self._connection.address) {
      digits = self._connection.address().address.split('.')
    }
    if (digits.length === 4) {
      cmd[3] = digits[3]
      cmd[4] = digits[2]
      cmd[5] = digits[1]
      cmd[6] = digits[0]
    } else {
      cmd[3] = 1
      cmd[4] = 0
      cmd[5] = 0
      cmd[6] = 127
    }

    // Byte 7: RF Signal strength
    // - 0x00, no signal
    // - 0x01, weak
    // - 0x02, low
    // - 0x03, medium
    // - 0x04, strong
    // - 0xFF, RF is not supported
    cmd[7] = 0xFF

    // Byte 8: Router status
    // - 0x00, wireless router is connected
    // - 0x01, wireless router not connected
    // - 0x02, connecting to a wireless router
    // - 0x03, password verification error
    // - 0x04, no wireless router found
    // - 0x05, IP cannot be obtained
    // - 0x06, wireless unstable
    //  - 0xFF, WI-FI failure
    cmd[8] = 0x00

    // Byte 9: Cloud service connection status:
    // - 0x00, connected to the cloud service center
    // - 0x01, not connected to the cloud service center
    // - 0x02, unstable internet connection
    // - 0x03, domain name resolution error
    // - 0x04, cloud service connection refused
    // - 0x05, cloud service maintenance
    // - 0xFF, cloud service failure
    cmd[9] = 0x01

    // Byte 10: Direct LAN connection status
    // - 0x00: No connection/connection has been disconnected
    // - 0x01: connected/connected with mobile terminal
    cmd[10] = 0x01

    // Byte 11 Number of TCP connections between the module and the mobile terminal
    cmd[11] = 0x01

    // Byte 12 - 19 Reserved
    cmd[12] = 0x00
    cmd[13] = 0x00
    cmd[14] = 0x00
    cmd[15] = 0x00
    cmd[16] = 0x00
    cmd[17] = 0x00
    cmd[18] = 0x00
    cmd[19] = 0x00

    cmd = createCommand(cmd, 0x0D, 0xAC, false)

    return new Promise((resolve, reject) => {
      self._request(cmd, 'sendNetworkStatusNotification')
        .then(response => {
          resolve(response)
        })
        .catch(error => {
          reject(error)
        })
    })
  }

  setStatus (properties = {}, retry = false, emitUpdates = true) {
    var self = this
    let cmd = Buffer.alloc(25)

    // Copy the current status
    const status = { ...self.status }

    // Enabe beep by default
    status.beep = true

    return new Promise((resolve, reject) => {
      const mode = {
        auto: 1,
        cool: 2,
        dry: 3,
        heat: 4,
        fanonly: 5
      }

      const fanSpeed = {
        auto: 102,
        silent: 20,
        low: 40,
        medium: 60,
        high: 80
      }

      for (const property in properties) {
        switch (property) {
          case 'beep':
            status.beep = properties[property] === true
            break

          case 'fanSpeed':

            if (!fanSpeed[properties[property]]) {
              return reject(new errors.OutOfRangeError('fanSpeed must be one of: auto, silent, low, medium or high'))
            }

            status.fanSpeed = fanSpeed[properties[property]]
            break

          case 'frostProtectionModeActive': // Requires capability frostProtectionMode and only available when mode is heat
            if (status.mode !== 'heat' && properties.mode !== 'heat') {
              return reject(new errors.OutOfRangeError('frostProtection capability is only available in heat mode'))
            }

            status.frostProtectionModeActive = properties[property] === true
            break

          case 'horizontalSwingActive':
            status.horizontalSwingActive = properties[property] === true
            break

          case 'mode':
            // FIXME: Only allow modes that are available according to the capabilities
            if (!mode[properties[property]]) {
              return reject(new errors.OutOfRangeError('Mode must be one of: auto, cool, dry, heat or fanonly'))
            }

            status.mode = mode[properties[property]]
            break

          case 'powerOn':
            status.powerOn = properties[property] === true
            break

          case 'setpoint':
            if ((status.temperatureUnit === 0 || properties.temperatureUnit === 'celcius') &&
              (properties[property] < 16 || properties[property] > 31)) {
              return reject(new errors.OutOfRangeError('The setpoint must be between 16 - 31°C'))
            }

            if (status.temperatureUnit === 1 || properties.temperatureUnit === 'fahrenheit') {
              if (properties[property] < 60 || properties[property] > 87) {
                return reject(new errors.OutOfRangeError('The setpoint must be between 60 - 87°F'))
              }
            }

            status.setpoint = properties[property]
            break

          case 'sleepModeActive':
            status.sleepModeActive = properties[property] === true
            break

          case 'temperatureUnit':
            if (properties[property] !== 'fahrenheit' && properties[property] !== 'celcius') {
              return reject(new errors.OutOfRangeError('The temperatureUnit must either be fahrenheit or celcius'))
            }

            status.temperatureUnit = properties[property] === 'fahrenheit' ? 0x01 : 0x00
            break

          case 'turboModeActive': // Requires capability strongCool and/or strongHeat
            status.turboModeActive = properties[property] === true
            break

          case 'verticalSwingActive':
            status.verticalSwingActive = properties[property] === true
            break

          default:
            return reject(new errors.OutOfRangeError(`Unsupported property to be set (${property})`))
        }
      }

      cmd[0] = 0x40

      // ABCDEFGH:
      // A: 0x00
      // B: beep
      // C: fastCheckActive
      // D: timerMode
      // E: childSleepMode
      // F: resumeActive
      // G: remoteControlMode (0: remote control, 1: PC)
      // H: powerOn
      status.remoteControlMode = 1
      cmd[1] = (status.beep ? 0x40 : 0x00) | (status.fastCheckActive ? 0x20 : 0x00) |
        (status.timerMode ? 0x10 : 0x00) | (status.childSleepMode ? 0x08 : 0x00) |
          (status.resume ? 0x04 : 0x00) | (status.remoteControlMode ? 0x02 : 0x00) |
            (status.powerOn ? 0x01 : 0x00)

      // AAABCCCC
      // A: mode
      // B: setpoint decimal (0.5)
      // C: setpoint
      let setpoint = status.setpoint
      if (setpoint > 60) {
        // Convert Fahrenheit to Celcius
        setpoint = Math.ceil((setpoint - 32) / 1.8 * 2) / 2
      }

      cmd[2] = (status.mode << 5) | (setpoint % 1 ? 0x10 : 0x00) | Math.floor(setpoint - 16)

      // ABBBBBBB
      // A: timerEffe?
      // B: fanSpeed
      cmd[3] = status.fanSpeed

      // ABBBBBCC
      // A: onTimerActive
      // B: onTimerHours
      // C: onTimerMinutes bits 0/1
      cmd[4] = (status.onTimerActive ? 0x80 : 0x00) | (status.onTimerHours << 2) | (status.onTimerMinutes & 0x03)

      // ABBBBBCC
      // A: offTimerActive
      // B: offTimerHours
      // C: offTimerMinutes bits 0/1
      cmd[5] = (status.offTimerActive ? 0x80 : 0x00) | (status.offTimerHours << 2) | (status.offTimerMinutes & 0x03)

      // AAAABBBB
      // A: timerOffMinutes bits 2-6
      // B: timerOffMinutes bits 2-6
      cmd[6] = ((status.timerOnMinutes & 0x2C) << 2) | (status.timerOffMinutes & 0x2C >> 2)

      // Nethome Plus: When timers are off, byte 4 & 5 are 0x7F and byte 6 is 0xFF

      // AAAABBCC
      // A: 0x03
      // B: horizontalSwingActive ? 0x0C : 0
      // C: verticalSwingActive ? 0x03 : 0
      cmd[7] = 0x30 | (status.horizontalSwingActive ? 0x0C : 0x00) | (status.verticalSwingActive ? 0x03 : 0x00)

      // ABCDEFGG
      // A: personalFeeling
      // B: powerSave
      // C: strong (=turboModeActive)
      // D: lowFrequencyFan
      // E: save
      // F: alarmSleep
      // G: cosySleepMode
      cmd[8] = (status.personalFeeling ? 0x80 : 0x00) | (status.powerSave ? 0x40 : 0x00) |
        (status.turboModeActive ? 0x20 : 0x00) | (status.lowFrequencyFan ? 0x10 : 0x00) |
          (status.save ? 0x08 : 0x00) | (status.alarmSleep ? 0x04 : 0x00) |
            (status.cosySleepMode & 0x03)

      // ABCDEFGH
      // A: ecoModeActive
      // B: changeCosySleepMode
      // C: purifyingModeActive (cleanUp)
      // D: ptcButton
      // E: ptcHeaterActive
      // F: dryClean
      // G: naturalWindModeActive (exchangeAir)
      // H: wiseEye
      cmd[9] = (status.ecoModeActive ? 0x80 : 0x00) | (status.purifyingModeActive ? 0x40 : 0x00) |
        (status.ptcHeaterActive ? 0x08 : 0x00) | (status.dryClean ? 0x04 : 0x00) |
          (status.naturalWindActive ? 0x02 : 0x00) | (status.wiseEye ? 0x01 : 0x00)

      // ABCDEFGH
      // A: cleanFanTime
      // B: dustFull
      // C: peakValleyMode
      // D: nightLight
      // E: catchCold
      // F: temperatureUnit (fahrenheit / celcius)
      // G: turboModeActive
      // H: sleepModeActive
      cmd[10] = (status.dustFull ? 0x40 : 0x00) | (status.peakValleyMode ? 0x20 : 0x00) |
        (status.nightLight ? 0x10 : 0x00) | (status.catchCold ? 0x08 : 0x00) |
        (status.temperatureUnit ? 0x04 : 0x00) | (status.turboModeActive ? 0x02 : 0x00) |
        (status.sleepModeActive ? 0x01 : 0x00)

      // AAAABBBB
      // A: sleepCurveTempPhase2
      // B: sleepCurveTempPhase1
      cmd[11] = 0x00

      // AAAABBBB
      // A: sleepCurveTempPhase4
      // B: sleepCurveTempPhase3
      cmd[12] = 0x00

      // AAAABBBB
      // A: sleepCurveTempPhase6
      // B: sleepCurveTempPhase5
      cmd[13] = 0x00

      // AAAABBBB
      // A: sleepCurveTempPhase8
      // B: sleepCurveTempPhase7
      cmd[14] = 0x00

      // ???this.naturalWindModeActive ? 0x40 : 0x00 (naturalFan)
      // AAAABBBB
      // A: sleepCurveTempPhase10
      // B: sleepCurveTempPhase9
      cmd[15] = 0x00

      // ABCDEFGH
      // A: tempUnitPhase8
      // B: tempUnitPhase7
      // C: tempUnitPhase6
      // D: tempUnitPhase5
      // E: tempUnitPhase4
      // F: tempUnitPhase3
      // G: tempUnitPhase2
      // H: tempUnitPhase1
      cmd[16] = 0x00

      // ABCDEEEE
      // A: pmv
      // B: natWind
      // C: tenHours
      // D: nightHours
      // E: sleepTime
      cmd[17] = 0x00

      // AAABBBB
      // A: pmv
      // B: temperature
      cmd[18] = 0x00

      // ABBBBBBB
      // A: humidityControlOnly
      // B: humidity
      cmd[19] = 0x00

      // ABBBBBBB
      // A: leftright
      // B: speed
      cmd[20] = 0x00

      // ABCCCCCD
      // A: frostProtectionModeActive
      // B: twoControl (double_temp)
      // C: temp (setExpand)
      // D: tempDot (setExpand_dot)
      cmd[21] = status.frostProtectionModeActive ? 0x80 : 0x00

      // AAAAAAAA
      // A: 0x00
      cmd[22] = 0x00

      // TODO: Setting byte 23 results for some commands in error 9999
      if (++self._messageId === 256) {
        self._messageId = 1
      }
      // cmd[23] = self._messageId

      cmd = createCommand(cmd, 0x02)

      // Send the command
      self._request(cmd, 'setStatus', retry)
        .then(response => {
          const parsedData = parse(response)

          // Update in-memory state
          const updates = self._updateStatus(parsedData)

          if (emitUpdates && Object.keys(updates).length) {
            self.emit('status-update', reporter(updates))
          }

          resolve(reporter(parsedData))
        })
        .catch(error => {
          reject(error)
        })
    })
  }
}
